jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -740,16 +753,21 @@ static void checkClassModuleVisibility(int accessMode, Module accessModule, Clas
 				if ((UNCONDITIONAL & accessMode) == UNCONDITIONAL) {
 					/* publicLookup objects can see all unconditionally exported packages. */
 					if (!targetModule.isExported(targetClassPackageName)) {
-						/*[MSG "K0587", "'{0}' no access to: '{1}'"]*/
-						throw new IllegalAccessException(Msg.getString("K0587", accessModule.getName(), targetClassPackageName)); //$NON-NLS-1$
+						/*[MSG "K0676", "Module '{0}' no access to: package '{1}' which is not exported by module '{2}'"]*/
+						throw new IllegalAccessException(Msg.getString("K0587", getModuleName(accessModule), targetClassPackageName, getModuleName(targetModule))); //$NON-NLS-1$
 					}
 				} else if (!(
-						Objects.equals(accessModule, targetModule)
-						|| (targetModule.isExported(targetClassPackageName, accessModule) && accessModule.canRead(targetModule)))
-						) {
-					/*[MSG "K0587", "'{0}' no access to: '{1}'"]*/
-					String message = Msg.getString("K0587", accessModule.getName(), targetClassPackageName); //$NON-NLS-1$
-					throw new IllegalAccessException(message);
+					Objects.equals(accessModule, targetModule)
+					|| (targetModule.isExported(targetClassPackageName, accessModule) && accessModule.canRead(targetModule)))
+				) {
+					if (!targetModule.isExported(targetClassPackageName, accessModule)) {
+						/*[MSG "K0677", "Module '{0}' no access to: package '{1}' which is not exported by module '{2}' to module '{0}'"]*/
+						throw new IllegalAccessException(Msg.getString("K0677", getModuleName(accessModule), targetClassPackageName, getModuleName(targetModule))); //$NON-NLS-1$
+					} else {
```

Comment:
@DanHeidinga, thanks. This `if-else` in question attempts to differentiate two cases: 
```
1. !targetModule.isExported(targetClassPackageName, accessModule)
```
Error message for this case: Module `accessModule` no access to: package `targetClassPackageName` which is not exported by module `targetModule` to module `accessModule`.
```
2. !accessModule.canRead(targetModule)
```
Error message for this case: Module `accessModule` no access to: package `targetClassPackageName` because module `accessModule` can't read module `targetModule`.

These two cases are different because `targetModule.isExported(targetClassPackageName, accessModule)` doesn't check if `accessModule.canRead(targetModule)` as per spec https://docs.oracle.com/javase/10/docs/api/java/lang/Module.html#isExported(java.lang.String,java.lang.Module).

Reply:
The code organization I was suggesting was more like:
```java
if (!targetModule.isExported(targetClassPackageName, accessModule)) {
	/*[MSG "K0677", "Module '{0}' no access to: package '{1}' which is not exported by module '{2}' to module '{0}'"]*/
	throw new IllegalAccessException(Msg.getString("K0677", getModuleName(accessModule), targetClassPackageName, getModuleName(targetModule))); //$NON-NLS-1$
} 
// here accessModule.canRead(targetModule) must be false
/*[MSG "K0679", "Module '{0}' no access to: package '{1}' because module '{0}' can't read module '{2}'"]*/
throw new IllegalAccessException(Msg.getString("K0679", getModuleName(accessModule), targetClassPackageName, getModuleName(targetModule))); //$NON-NLS-1$
 					
```
As the if side of the branch unconditionally throws.

It makes it slightly easier to see that every path through this code will throw an exception.

Reply:
Aha, got it. Fixed.